server/src/openfga-yaml-schema.ts

```diff
@@ -1,4 +1,306 @@
+    return false;
+  }
+  return true;
+}
+
+// Validation for Types
+
+function validateTypes(tuple: TupleKey, types: string[], instancePath: string): boolean {
+  const errors = [];
+
+  const user = tuple.user.split(":")[0];
+
+  // Ensure valid type of user
+  if (!types.includes(user)) {
+    errors.push(invalidType(user, types, instancePath + "/user"));
+  }
+
+  const object = tuple.object.split(":")[0];
+
+  // Ensure valid type of object
+  if (!types.includes(object)) {
+    errors.push(invalidType(object, types, instancePath + "/object"));
+  }
+
+  // Report all errors
+  if (errors.length) {
+    validateTuple.errors?.push(...errors);
+    return false;
+  }
+  return true;
+}
+
+// Validate Relation
+function validateRelation(
+  tuple: TupleKey,
+  typeDefs: TypeDefinition[],
+  instancePath: string): boolean {
+  const errors = [];
+
+  // Check if relation exists on given type
+  let doesExistOnType = false;
+  if (tuple.user.includes("#")) {
+    const user = tuple.user.split(":")[0];
+    const userRelation = tuple.user.split("#")[1];
+    for (const typeDef of typeDefs) {
+      if (typeDef && typeDef.type === user && typeDef.relations?.[userRelation]) {
+        doesExistOnType = true;
+        break;
+      }
+    }
+
+    if (!doesExistOnType) {
+      errors.push(relationMustExistOnType(userRelation, user, instancePath + "/user"));
+    }
+  }
+
+  // Check if relation exists on given object
+  const objectType = tuple.object.split(":")[0];
+  let doesExistOnObject = false;
+  for (const typeDef of typeDefs) {
+    if (typeDef && typeDef.type === objectType && typeDef.relations?.[tuple.relation]) {
+      doesExistOnObject = true;
+      break;
+    }
+  }
+  if (!doesExistOnObject) {
+    errors.push(relationMustExistOnType(tuple.relation, objectType, instancePath + "/relation"));
+  }
+
+  if (errors.length) {
+    validateTuple.errors?.push(...errors);
+    return false;
+  }
+  return true;
+}
+
+type MappedTuple = {
+  type: string,
+  relation?: string | undefined,
+  wildcard?: object | undefined,
+  condition?: RelationshipCondition,
+}
+
+function mapTuple(tuple: TupleKey): MappedTuple {
+  return {
+    type: tuple.user.split(":")[0],
+    relation: tuple.user.includes("#") ? tuple.user.split("#")[1] : undefined,
+    wildcard: tuple.user.includes(":*") ? {} : undefined,
+    condition: tuple.condition,
+  };
+}
+
+function validateTypeRestrictions(
+  tuple: TupleKey,
+  typeDefs: TypeDefinition[],
+  conditions: {[key: string]: Condition} | undefined,
+  instancePath: string): boolean {
+
+    validateTuple.errors = validateTuple.errors || [];
+
+    const mappedTuple = mapTuple(tuple);
+    const object = tuple.object.split(":")[0];
+    const type = typeDefs.filter(t => t.type === object)[0];
+
+    // Find no direct relationships
+    if (type.metadata?.relations?.[tuple.relation].directly_related_user_types) {
+      const userTypes = type.metadata.relations[tuple.relation].directly_related_user_types;
```

Comment:
This section up until the validateConditionParams could be something like this

```js
function getRelationReferenceString(relationReference: RelationReference) {
  let relationReferenceString = relationReference.type;
  if (relationReference.wildcard) {
    relationReferenceString += ':*';
  } else if (relationReference.relation) {
    relationReferenceString += `#${relationReference.relation}`;
  }
  if (relationReference.condition) {
    relationReferenceString += ` with ${relationReference.condition}`;
  }
  return relationReferenceString;
}

const userTypes = type.metadata.relations[tuple.relation].directly_related_user_types;

if (!userTypes?.filter(
  userType => userType.type === mappedTuple.type // type matches
    && (!!mappedTuple.wildcard === !!userType.wildcard) // and the wildcard matches (either both true or both false)
    && userType.relation === mappedTuple.relation // and the relation matches
    && userType.condition === mappedTuple.condition // and the condition matches
  ).length) {
  validateTuple.errors.push(userNotTypeRestriction(getRelationReferenceString(mappedTuple), tuple, instancePath));
  return false;
}
```

Reply:
Done! Had to keep the if guard around `validateConditionParams` but great suggestion!